text/0000-cargo-report-future-incompat.md

```diff
@@ -0,0 +1,765 @@
+Example:
+
+```
+crates % cd unwary
+unwary % cargo build
+   Compiling brash v0.2.0
+   Compiling bold2 v0.1.0
+   Compiling unwary v0.2.0
+    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
+unwary %
+```
+
+Here, the unwary (sic) crate has updated its version of `brash`,
+switched to `bold2` (a fork of `bold`), and replaced its internal
+usage of `rash` with some local code, thus completely eliminating all
+current future-incompatibility lint triggers.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As noted above, we want to continue to suppress normal lint checks for
+upstream dependencies. Therefore, Cargo will continue to pass
+`--cap-lints=allow` for non-path upstream dependencies.
+
+At the same time, we want to minimize disruption to existing users of Rust.
+Therefore, the behavior of flags that directly interact with lints, like
+`-Dwarnings`, will remain unchanged by this RFC.
+
+For example, in our running example of `unwary`:
+  * running `cargo rustc -- -Dwarnings` will behave the same as today
+    (the lints in the downstream `brash` dependency will be capped
+    and their warnings hidden), and
+  * running `RUSTFLAGS=-Dwarnings` will also behave the same as today
+    (the `rustc` invocation on the downstream `brash` depedency
+    will be invoked with `-Dwarnings` and thus the build of the `brash`
+    dependency will fail).
+
+However, the Rust compiler's behavior will change slightly. Even when
+`--cap-lints=allow` is turned on, we need Cargo to know when a
+future-incompatibilty lint is triggered.
+
+The division of responsbiilties between Cargo and the Rust compiler
+may be a little subtle:
+
+The responsibilities of the Rust compiler (`rustc`):
+
+  * `rustc` must differentiate future-incompatibility lints (c.f.
+     [PR #59658: "Minimum Lint Levels"][rust-lang/rust#59658]) from
+     other lints that are expected to remain as mere warnings forever.
+
+  * `rustc` will need to have some new mode of operation, which this
+    RFC will call the where it will check for instances of
+    future-incompatibility lints, *regardless* of whether
+    `--cap-lints=allow` is also set. This RFC calls this the
+    *future-incompatibility checking mode*.
+
+    * In the future-incompatibility checking mode of invocation,
+      `rustc` will also need to check for such lints *regardless* of
+      whether the code appears in the scope of an `#[allow(..)]`
+      attribute for the lint.
+
+    * In the future-incompatibility checking mode of invocation,
+      *emission* of the diagnostics themselves may still be silenced
+      as specified by `--cap-lints=allow` or `#[allow(..)]` attributes.
+
+    * That is, those flags and annotations should be interpreted by
+      `rustc` as silencing the diagnostic report, but *not* as
+      silencing the feedback about there existing some instance of the
+      lint triggering somewhere in the crate's source code.
```

Comment:
Right, so *if* the "feedback" is considered a warning in itself, then some might expect `-Dwarnings` to fail the build of the dependency and thus fail the entire build.  As I tried to suggest previously, this is an implementation choice. If you want to make `-Dwarnings` be _less lenient_ in this regard then it is today, then I think that _along with this change_ we need something like `-Dmy_warnings` or `-Dlocal_warnings` or some equivalent at the cargo flags level, that would be provided as an equivalent workaround for projects using `-Dwarnings` in CI today.  Otherwise, I think projects will practically need to either reinstate a local `deny(warnings)`, which is a pain point for iterative local development, or abandon warning checks in CI altogether.  And I don't think the later is consistent with your goals here.

Reply:
My intent is that the future-incompatibility report issued by cargo is itself not a warning. It is a report.

More specifically: my intent is that the observable behavior of `-Dwarnings` remain exactly the same with this RFC as it is without.

Does that address your concern?

Reply:
With the change requested, above yes!